### ISSUE TITLE
'Read commands must ignore leavers' issue under load

### ISSUE THREAD

**Issue by balladali** [keycloak/keycloak#20971]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

infinispan

### Describe the bug

We are using Keycloak in cluster mode with 2 instances (behind load balancer) and got exceptions during load testing of token introspection or exchange. These operations may fail with the following exceptions.

```
[2023-06-13 13:11:03,572] [INFO] [thread=jgroups-39,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=work] ISPN100010: Finished rebalance with members [my-comp-25105, my-comp-28516], topology id 16
[2023-06-13 13:11:28,327] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=authenticationSessions] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,396] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=sessions] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,420] [ERROR] [thread=jgroups-60,my-comp-25105] [caller=org.infinispan.interceptors.impl.InvocationContextInterceptor] [requestId=] ISPN000136: Error executing command TouchCommand on Cache 'sessions', writing keys []: org.infinispan.remoting.transport.jgroups.SuspectException: ISPN000400: Node my-comp-28516 was suspected
        at org.infinispan.remoting.transport.ResponseCollectors.remoteNodeSuspected(ResponseCollectors.java:31)
        at org.infinispan.remoting.transport.impl.SingleResponseCollector.targetNotFound(SingleResponseCollector.java:31)
        at org.infinispan.remoting.transport.impl.SingleResponseCollector.targetNotFound(SingleResponseCollector.java:17)
        at org.infinispan.remoting.transport.ValidSingleResponseCollector.addResponse(ValidSingleResponseCollector.java:23)
        at org.infinispan.remoting.transport.impl.SingleTargetRequest.addResponse(SingleTargetRequest.java:75)
        at org.infinispan.remoting.transport.impl.SingleTargetRequest.onResponse(SingleTargetRequest.java:45)
        at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1578)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1480)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1685)
        at org.jgroups.JChannel.up(JChannel.java:733)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:921)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:138)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:845)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:226)
        at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1052)
        at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:794)
        at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:776)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:425)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:658)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:188)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:294)
        at org.jgroups.stack.Protocol.up(Protocol.java:314)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1178)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:100)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

[2023-06-13 13:11:28,443] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=clientSessions] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,471] [ERROR] [thread=jgroups-60,my-comp-25105] [caller=org.infinispan.interceptors.impl.InvocationContextInterceptor] [requestId=] ISPN000136: Error executing command GetKeyValueCommand on Cache 'sessions', writing keys []: java.lang.IllegalStateException: Read commands must ignore leavers
        at org.infinispan.interceptors.impl.BaseStateTransferInterceptor.handleExceptionOnReadCommandReturn(BaseStateTransferInterceptor.java:161)
        at org.infinispan.interceptors.impl.BaseStateTransferInterceptor.handleReadCommandReturn(BaseStateTransferInterceptor.java:204)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.invokeQueuedHandlers(QueueAsyncInvocationStage.java:124)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:87)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:33)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.invokeQueuedHandlers(QueueAsyncInvocationStage.java:114)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:87)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:33)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
        at org.infinispan.remoting.transport.AbstractRequest.completeExceptionally(AbstractRequest.java:75)
        at org.infinispan.remoting.transport.impl.SingleTargetRequest.onResponse(SingleTargetRequest.java:49)
        at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1578)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1480)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1685)
        at org.jgroups.JChannel.up(JChannel.java:733)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:921)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:138)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:845)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:226)
        at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1052)
        at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:794)
        at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:776)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:425)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:658)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:188)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:294)
        at org.jgroups.stack.Protocol.up(Protocol.java:314)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1178)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:100)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

[2023-06-13 13:11:28,511] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=work] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,537] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=org.infinispan.CLIENT_SERVER_TX_TABLE] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,666] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=loginFailures] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,534] [WARN] [thread=executor-thread-62] [caller=org.keycloak.events] [requestId=] type=INTROSPECT_TOKEN_ERROR, realmId=7913950a-e6d6-407d-8002-d895bb4e678f, clientId=test-introspect, userId=null, ipAddress=127.0.0.1, error=invalid_request, detail='Failed to introspect token.', client_auth_method=client-secret
[2023-06-13 13:11:28,701] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=offlineClientSessions] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,741] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=org.infinispan.CONFIG] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,886] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=org.infinispan.COUNTER] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,942] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=offlineSessions] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:28,970] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] [Context=actionTokens] ISPN100008: Updating cache members list [my-comp-25105], topology id 17
[2023-06-13 13:11:29,035] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] ISPN000094: Received new cluster view for channel ISPN: [my-comp-25105|6] (1) [my-comp-25105]
[2023-06-13 13:11:29,047] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] ISPN100001: Node my-comp-28516 left the cluster
[2023-06-13 13:11:29,055] [INFO] [thread=jgroups-17,my-comp-25105] [caller=org.infinispan.CLUSTER] [requestId=] ISPN100001: Node my-comp-28516 left the cluster
```

It happens when many threads (70-100) per second are introspecting or exchanging access token and one of Keycloak instance suspends by some reason. In this case token introspection or exchanging returns error response.

The important point here is that one instance is suspending and nodes are rebalancing during token introspection/exchange. Looks like alive node is trying to read cache from node that is leaving at this moment.

### Version

21.1.1

### Expected behavior

Successful token introspection or exchange.

### Actual behavior

Failed token introspection (400 Bad Request) or exchange (401 Unauthorized)

### How to Reproduce?

1. Start Keycloak in cluster with 2 instances
2. Get access token
3. Introspect or exchange access token under load (~70-100 requests per second)
4. Shut down one of 2 Keycloak instances

**Comment by martin-kanis** [comment]
@balladali Thanks for reporting the issue.
Reproduced with "keycloak-benchmark", latest nightly image.
Scenarios:
```
./kcb.sh --server-url=... --realm-name=realm-0 --scenario=keycloak.scenario.authentication.AuthorizationCode --users-per-sec=150 --measurement=300 --logout-percentage=0 --users-per-realm=10000 --ramp-up=100
./kcb.sh --server-url=... --realm-name=realm-0 --scenario=keycloak.scenario.authentication.AuthorizationCode --users-per-sec=80 --measurement=300 --logout-percentage=100 --users-per-realm=10000 --ramp-up=100

```

**Comment by wburns** [comment]
@martin-kanis Do you think you could run with a small change to ISPN to add a printStackTrace call at https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java#L151 which should show where the original exception came from

**Comment by martin-kanis** [comment]
> @martin-kanis Do you think you could run with a small change to ISPN to add a printStackTrace call at https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java#L151 which should show where the original exception came from

@wburns Yes, I will try to re-run the test with the proposed change. Do you want to also enable debug/trace log level and if so, for which component?

**Comment by wburns** [comment]
I don't think tracing in other components would be required. We just need to see where the SuspectException was raised from and fix that part at this point.

**Comment by martin-kanis** [comment]
oh sorry! Here it is:
```
org.infinispan.remoting.transport.jgroups.SuspectException: ISPN000400: Node keycloak-0-15376 was suspected
	at org.infinispan.remoting.transport.ResponseCollectors.remoteNodeSuspected(ResponseCollectors.java:31)
	at org.infinispan.remoting.transport.impl.SingleResponseCollector.targetNotFound(SingleResponseCollector.java:31)
	at org.infinispan.remoting.transport.impl.SingleResponseCollector.targetNotFound(SingleResponseCollector.java:17)
	at org.infinispan.remoting.transport.ValidSingleResponseCollector.addResponse(ValidSingleResponseCollector.java:23)
	at org.infinispan.remoting.transport.impl.SingleTargetRequest.addResponse(SingleTargetRequest.java:75)
	at org.infinispan.remoting.transport.impl.SingleTargetRequest.onResponse(SingleTargetRequest.java:45)
	at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1578)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1480)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.lambda$up$1(JGroupsTransport.java:1699)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1691)
	at org.jgroups.JChannel.up(JChannel.java:749)
	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:924)
	at org.jgroups.protocols.FRAG2.up(FRAG2.java:160)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:319)
	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:859)
	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:246)
	at org.jgroups.protocols.UNICAST3.deliverBatch(UNICAST3.java:1077)
	at org.jgroups.protocols.UNICAST3.handleBatchReceived(UNICAST3.java:854)
	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:521)
	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:700)
	at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:119)
	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:193)
	at org.jgroups.stack.Protocol.up(Protocol.java:346)
	at org.jgroups.protocols.MERGE3.up(MERGE3.java:288)
	at org.jgroups.protocols.Discovery.up(Discovery.java:314)
	at org.jgroups.protocols.RED.up(RED.java:119)
	at org.jgroups.protocols.TP.passBatchUp(TP.java:1204)
	at org.jgroups.util.SubmitToThreadPool$BatchHandler.passBatchUp(SubmitToThreadPool.java:154)
	at org.jgroups.util.SubmitToThreadPool$BatchHandler.run(SubmitToThreadPool.java:150)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by wburns** [comment]
Thanks, I will take a closer look this week.

**Comment by martin-kanis** [comment]
@jabolina Thanks for looking into this. Would it be possible to backport it to the 14.x branch? Keycloak needs to stay on 14 for now and I'm not sure we can try with 15 at this point without breaking anything.

**Comment by jabolina** [comment]
@martin-kanis We just included a backport to 14.

**Comment by martin-kanis** [comment]
@jabolina Thanks. I ran keycloak-benchmark with 14.0.x branch, and I didn't see any `Read commands must ignore leavers` errors. 

So the issue should be fixed by ISPN upgrade https://github.com/keycloak/keycloak/issues/22855

**Comment by martin-kanis** [comment]
Fixed by #22887
